[DOC-13921]: XDCR: Tuners to Throttle DCP Control Flow & Async Events Channel Length (#4107)
couchbase/docs-server: modules/xdcr-reference/pages/xdcr-lowering-memory-footprint.adoc

= Lowering the Memory Footprint in XDCR
:description: Using new global settings (introduced in Couchbase Server 8.1) to lower the memory footprint of Cross Data Center Replication.
:stem: asciimath

[abstract]
{description}

> XDCR has 2 global replication settings which can be used to lower the memory footprint of a running replication:

Reviewer: >  which can be used 

Grammar nit: use active voice.

```suggestion
XDCR has 2 global replication settings that you can adjust to lower the memory footprint of a running replication:
```

dcpFlowControlThrottle::
This is a percentage applied to the default value of both: 
+
--

> * The `connection_buffer_size` set by XDCR replication for the DCP connection to KV (stem:[1024 times 1024] bytes by default)

Reviewer 2: `stem:[1024 times 1024]` - I think regular font will work for this? just `1024 * 1024`

Author: You're right: using stem: here is overkill.

* Length of the channel used by XDCR replication to buffer incoming items (via the DCP connection) from KV (20000 by default)
--
+
Effective value for both the above parameters becomes: 
+
--

> stem:[("<default value> " times " dcpFlowControlThrottle") / 100]

Reviewer: In my browser (Firefox on Linux), the numerator of this equation does not appear. All I see is the lower half of the equation. I know that the math formatting can sometimes be a bit finicky in preview. So, if it shows up for you, it's likely OK.

--
+
Reducing this setting from its default value of 100 causes a decrease in the max number of documents that can fit into XDCR's memory buffer.
 This limitation on the size of the buffer results in a drop in the replication throughput, thereby reducing the memory footprint of the replication.

componentEventsChanLength::
This sets the default length of the channels being used by event-listeners to buffer various internal events before processing.
+
Reducing this setting from its default value of 10,000 lowers the number of buffer slots available to each of the event-listeners, thereby reducing the memory allocated. 
However, it may also cause a drop in replication throughput.

> NOTE: Changing these replication settings for an ongoing replication causes it to restart from the last checkpoint sequence number

Reviewer 2: "the last checkpoint sequence number" can be "the last checkpoint's sequence number" or "the last checkpointed sequence number"

Author: fixed

so that the new values can take effect.

== Monitoring the Health of a Throttled Replication

> Lowering either of these settings may adversely affect the throughput of the replication, and care must be taken to not over-throttle the replication.

Reviewer: I would break this into two sentences:

```suggestion
Lowering either of these settings may adversely affect the throughput of the replication.
Be careful to not over-throttle the replication.
```

The health of the replication in response to the throttling can be measured via the following Prometheus metrics expression:

[source, text]
----

> clamp_min(xdcr_changes_left_total\{pipelineType="Main", sourceBucketName="<A>", targetBucketName="<B>", targetClusterUUID="<C>"} - ignoring(name) rate(xdcr_docs_processed_total[1m]), 0)

Reviewer 2: typo - the back-slash in `xdcr_changes_left_total\{pipelineType="Main"`  is extra, it may cause the expression to break

Author: fixed. Thanks.
